### PR TITLE
BUG: Chained mutate operations give wrong results

### LINKED ISSUE #1799: BUG: Chained mutate operations give wrong results
Mutate function is giving wrong output. Please check below sample example. 

```import ibis 
table_expr = ibis.table([('a', 'string'), ('b', 'string')]) 
table_expr1 = table_expr['b'].fillna('Short Term') 
table_expr = table_expr.mutate(b = table_expr1) 
table_expr2 = table_expr['a'].fillna('Short Term') 
table_expr = table_expr.mutate(a = table_expr2) 
print(ibis.impala.compile(table_expr)) 
```
For the above program output is :
`SELECT isnull(`a`, 'Short Term') AS `a`, `b`
FROM t0`

instead of 
`SELECT isnull(`a`, 'Short Term') AS `a`, isnull(`b`, 'Short Term') AS `b`
FROM t0`

Let me know if anything is wrong in my understanding.

### PR DESCRIPTION
**Pull request by cpcloud**
Closes #1799

This was kind of a nasty one. The query flattening optimization was far too
aggressive.

@toryhaavik Do you feel comfortable reviewing this?

One downside to this fix is that simple, single expression columns are not flattened into their subquery anymore. This will be addressed in the future after the compiler is refactored.

### CODE REVIEW

File: ibis/expr/analysis.py [hunk 1]
```diff
@@ -343,26 +345,30 @@ def _lift_TableColumn(self, expr, block=None):
 
         if isinstance(root, ops.Selection):
             can_lift = False
+            all_simple_columns = all(
+                isinstance(sel.op(), ops.TableColumn)
+                and sel.op().name == sel.get_name()
+                for sel in root.selections
+                if isinstance(sel, ir.ValueExpr)
+                if sel.has_name()
+            )
```
Comment (cpcloud) [inline ibis/expr/analysis.py]: This is the fix for the issue. This predicate prevents lifting columns that are projected out of `Selections` with anything but simple column accesses.

**Comment by toryhaavik** [comment]
Sure, I'll review

File: ibis/expr/api.py [hunk 2]
```diff
@@ -3360,43 +3360,29 @@ def mutate(table, exprs=None, **mutations):
...
-    if exprs is None:
-        exprs = []
-    else:
-        exprs = util.promote_list(exprs)
+    exprs = [] if exprs is None else util.promote_list(exprs)
+    exprs.extend(
+        (expr(table) if util.is_function(expr) else as_value_expr(expr)).name(
+            name
+        )
+        for name, expr in sorted(mutations.items(), key=operator.itemgetter(0))
```
Comment (cpcloud) [inline ibis/expr/api.py]: This sorting can likely be removed.

**Comment by toryhaavik** [comment]
This makes sense to me, let's merge it.

**Comment by cpcloud** [comment]
Merging.

**Pull request merged**